stan/math/opencl/rev/copy.hpp

```diff
@@ -16,6 +16,74 @@
 namespace stan {
 namespace math {
 
+namespace internal {
+template <typename T_arg_adj, require_eigen_t<T_arg_adj>* = nullptr>
+class op_copy_to_cl_vari final
+    : public vari_value<matrix_cl<value_type_t<T_arg_adj>>> {
+  T_arg_adj arg_adj_;
+
+ public:
+  template <typename T_arg_val, require_eigen_t<T_arg_val>* = nullptr,
+            require_vt_same<T_arg_val, T_arg_adj>* = nullptr>
+  explicit op_copy_to_cl_vari(const T_arg_val& val, T_arg_adj adj)
+      : vari_value<matrix_cl<value_type_t<T_arg_adj>>>(to_matrix_cl(val)),
+        arg_adj_(adj) {}
+
+  virtual void chain() {
+    arg_adj_ += from_matrix_cl<std::decay_t<T_arg_adj>::RowsAtCompileTime,
+                               std::decay_t<T_arg_adj>::ColsAtCompileTime>(
+        this->adj_);
+  }
+};
+}  // namespace internal
+
+/** \ingroup opencl
+ * Copies the source var containing Eigen matrices to destination var that has
+ * data stored on the OpenCL device.
+ *
+ * @tparam T type of the Eigen matrix
+ * @param a source Eigen matrix
+ * @return var with a copy of the data on the OpenCL device
+ */
+template <typename T>
+inline var_value<matrix_cl<value_type_t<T>>> to_matrix_cl(var_value<T>& a) {
+  return new internal::op_copy_to_cl_vari<decltype(a.adj())>(a.val(), a.adj());
+}
+
+namespace internal {
+template <typename T, int Rows, int Cols,
+          require_all_kernel_expressions_t<T>* = nullptr>
+class op_copy_from_cl_vari final
+    : public vari_value<Eigen::Matrix<value_type_t<T>, Rows, Cols>> {
+  vari_value<T>& a_;
+
+ public:
+  explicit op_copy_from_cl_vari(vari_value<T>& a)
+      : vari_value<Eigen::Matrix<value_type_t<T>, Rows, Cols>>(
+            from_matrix_cl<Rows, Cols>(a.val_)),
+        a_(a) {}
+
+  virtual void chain() { a_.adj_ = a_.adj_ + to_matrix_cl(this->adj_); }
```

Comment:
Do we not have a `+=` operator for `matrix_cl`?

Reply:
nope